src/Disks/ObjectStorages/DiskObjectStorage.cpp

```diff
@@ -177,12 +188,13 @@ void DiskObjectStorage::copyFile( /// NOLINT
     const ReadSettings & read_settings,
     const WriteSettings & write_settings)
 {
-    if (this == &to_disk)
+    if (getDataSourceDescription() == to_disk.getDataSourceDescription())
```

Comment:
What has to be equal here to perform server side?
Could it be
- bucket?
- credentials?

What client should perform such requst?

PS: Types `s3` and `s3_plain` actually differ only in the way how CH maps files to blob keys.

Reply:
>What has to be equal here to perform server side?

In fact, as I understood from  the doc([CopyObjectS3](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html)) and the current implementation of [`backup s3_policy to s3(...)`](https://github.com/ClickHouse/ClickHouse/blame/4ed4fbe74df0f48382bad2c55c883b763e5c869f/src/Backups/BackupIO_S3.cpp#L219) there are no any limitations on s3 side.  We limited it to the description of the data source so as not to handle scenarios when metadata_storage does not match. IMO, this may be more complicated than in the case with equal metadata_storage type. What is your opinion? Can we use server side with `s3` -> `s3_plain` scenario?

>What client should perform such requst?

Do you mean client for source s3 or destination? Seems like client must be to destination object_storage similar like [here](https://github.com/ClickHouse/ClickHouse/blob/4ed4fbe74df0f48382bad2c55c883b763e5c869f/src/Backups/BackupIO_S3.cpp#L197). But right now requests for copy object are sending to source s3, will fix it.

Reply:
> Do you mean client for source s3 or destination? Seems like client must be to destination object_storage similar like [here](https://github.com/ClickHouse/ClickHouse/blob/4ed4fbe74df0f48382bad2c55c883b763e5c869f/src/Backups/BackupIO_S3.cpp#L197). But right now requests for copy object are sending to source s3, will fix it.

Yes I agree, it is natural to get the dst-client for such operation. Because that client needs to have only additional RO access to other disk. Otherwise the client needs to have write access to other disk.

Reply:
> Can we use server side with s3 -> s3_plain scenario?
You do not have to solve this actually here. I dont know how to solve it either. That entities are pure logical. That decision depends on its logic. If I have to do it, I would revise the reasons why that types of disks even exists.

Reply:
@CheSema Hi! Could you please check the PR?

Reply:
@CheSema ping

Reply:
>Do we perform server side copy for different buckets?

Yes. as I can see we're already performing native copy in scenario: `backup .. to s3`([the test](https://github.com/ClickHouse/ClickHouse/blob/e04416b48f20037ec2da5deb411254d3c993592e/tests/integration/test_backup_restore_s3/test.py#L317))